File: crates/prolly/bindings/ruby/examples/batch_build.rb
# frozen_string_literal: true

require 'tmpdir'

$LOAD_PATH.unshift(File.expand_path('../lib', __dir__))
require 'prolly'

def upsert(key, value)
  Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: key.b, value: value.b)
end

def delete(key)
  Prolly::MutationRecord.new(kind: Prolly::MutationKind::DELETE, key: key.b, value: nil)
end

def assert_equal(expected, actual, label = 'value')
  return if expected == actual

  raise "#{label}: expected #{expected.inspect}, got #{actual.inspect}"
end

def batch_build
  engine = Prolly::ProllyEngine.memory(Prolly.default_config)
  entries = (1..64).to_a.reverse.map do |idx|
    Prolly::EntryRecord.new(key: format('event/%04d', idx).b, value: "payload-#{idx}".b)
  end
  tree = engine.build_from_entries(entries)
  rows = engine.range(tree, 'event/'.b, 'event0'.b)
  stats = engine.collect_stats_json(tree).json

  assert_equal 64, rows.length, 'row count'
  assert_equal 'event/0001'.b, rows.first.key, 'first key'
  raise 'stats missing num_nodes' unless stats.include?('num_nodes')

  puts "batch_build: imported #{rows.length} events"
end

batch_build

File: crates/prolly/bindings/python/examples/batch_build.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from tempfile import TemporaryDirectory
import sys

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

import prolly


def b(value: str) -> bytes:
    return value.encode("utf-8")


def text(value: bytes | None) -> str | None:
    return None if value is None else value.decode("utf-8")


def assert_bytes(expected: bytes, actual: bytes | None, label: str) -> None:
    if actual != expected:
        raise AssertionError(f"{label}: expected {expected!r}, got {actual!r}")


def upsert(key: str, value: str | bytes) -> prolly.MutationRecord:
    payload = value if isinstance(value, bytes) else b(value)
    return prolly.MutationRecord(kind=prolly.MutationKind.UPSERT, key=b(key), value=payload)


def delete(key: str) -> prolly.MutationRecord:
    return prolly.MutationRecord(kind=prolly.MutationKind.DELETE, key=b(key), value=None)

def batch_build() -> None:
    engine = prolly.ProllyEngine.memory(prolly.default_config())
    entries = [
        prolly.EntryRecord(key=b(f"event/{idx:04d}"), value=b(f"payload-{idx}"))
        for idx in range(64, 0, -1)
    ]
    tree = engine.build_from_entries(entries)
    rows = engine.range(tree, b"event/", b"event0")
    stats = engine.collect_stats_json(tree).json

    assert len(rows) == 64
    assert rows[0].key == b"event/0001"
    assert "num_nodes" in stats

    print(f"batch_build: imported {len(rows)} events")


if __name__ == "__main__":
    batch_build()
